File: GildedRose/src/com/gildedrose/items/BackstagePassItem.java
package com.gildedrose.items;

public class BackstagePassItem extends ItemDecorator {
	
	// Constants
	private final int THRESHOLD_ONCE = 10;	
	private final int THRESHOLD_TWICE = 5;	
	private final int THRESHOLD_THRICE = 0;	

	// Constructors
	protected BackstagePassItem(Item itemToDecorate)
	{
		super(itemToDecorate);
	}

	// Protected methods
	@Override
	protected void updateQuality()
	{
		if (this.getSellIn() > THRESHOLD_ONCE)
		{
			this.increaseQualityOnce();
		}
		else if (this.getSellIn() > THRESHOLD_TWICE)
		{
			this.increaseQualityTwice();
		}
		else if (this.getSellIn() > THRESHOLD_THRICE)
		{
			this.increaseQualityThrice();
		}
		else  // this.isSellInEnded()
		{
			this.setQuality(this.MIN_QUALITY);
		}
	}

}

File: GildedRose.Net/Items/BackstagePassItem.cs
namespace GildedRose.Items
{
    public class BackstagePassItem : ItemDecorator
    {
		
		// Constants
		private const int THRESHOLD_ONCE = 10;	
		private const int THRESHOLD_TWICE = 5;	
		private const int THRESHOLD_THRICE = 0;	

		// Constructors
		internal BackstagePassItem(Item itemToDecorate) : base(itemToDecorate) { }
	
		// Protected methods
		protected override void UpdateQuality()
		{
			if (this.SellIn > THRESHOLD_ONCE)
			{
				this.IncreaseQualityOnce();
			}
			else if (this.SellIn > THRESHOLD_TWICE)
			{
				this.IncreaseQualityTwice();
			}
			else if (this.SellIn > THRESHOLD_THRICE)
			{
				this.IncreaseQualityThrice();
			}
			else  // this.IsSellInEnded()
			{
				this.Quality = MIN_QUALITY;
			}
		}
		
	}
}